samples/core/xgboost_training_cm/xgboost_training_cm.py

```diff
@@ -212,75 +215,90 @@ def xgb_train_pipeline(
-            eval_data,
-            target,
-            analyze_op.output,
-            output_template
-        ).apply(gcp.use_gcp_secret('user-gcp-sa'))
-
-        train_op = dataproc_train_op(
-            project,
-            region,
-            create_cluster_op.output,
-            transform_op.outputs['train'],
-            transform_op.outputs['eval'],
-            target,
-            analyze_op.output,
-            workers,
-            rounds,
-            output_template
-        ).apply(gcp.use_gcp_secret('user-gcp-sa'))
-
-        predict_op = dataproc_predict_op(
-            project,
-            region,
-            create_cluster_op.output,
-            transform_op.outputs['eval'],
-            train_op.output,
-            target,
-            analyze_op.output,
-            output_template
-        ).apply(gcp.use_gcp_secret('user-gcp-sa'))
-
-        confusion_matrix_task = confusion_matrix_op(
-            predict_op.output,
-            output_template
-        ).apply(gcp.use_gcp_secret('user-gcp-sa'))
+    output_template = str(output) + '/' + dsl.RUN_ID_PLACEHOLDER + '/data'
+
+    # Current GCP pyspark/spark op do not provide outputs as return values, instead,
+    # we need to use strings to pass the uri around.
+    analyze_output = output_template
+    transform_output_train = os.path.join(output_template, 'train', 'part-*')
+    transform_output_eval = os.path.join(output_template, 'eval', 'part-*')
+    train_output = os.path.join(output_template, 'train_output')
+    predict_output = os.path.join(output_template, 'predict_output')
+
+    with dsl.ExitHandler(exit_op=dataproc_delete_cluster_op(
+        project_id=project,
+        region=region,
+        name=cluster_name
+    )):
+        create_cluster_task = dataproc_create_cluster_op(
+            project_id=project,
+            region=region,
+            name=cluster_name,
+            initialization_actions=[
+              os.path.join(_PYSRC_PREFIX,
+                           'initialization_actions.sh'),
+            ],
+            image_version='1.2'
+        )
+
+        analyze_task = dataproc_analyze_op(
+            project=project,
+            region=region,
+            cluster_name=cluster_name,
+            schema=schema,
+            train_data=train_data,
+            output=output_template
+        ).after(create_cluster_task).set_display_name('Analyzer')
+
+        transform_task = dataproc_transform_op(
+            project=project,
+            region=region,
+            cluster_name=cluster_name,
+            train_data=train_data,
+            eval_data=eval_data,
+            target=target,
+            analysis=analyze_output,
+            output=output_template
+        ).after(analyze_task).set_display_name('Transformer')
+
+        train_task = dataproc_train_op(
+            project=project,
+            region=region,
+            cluster_name=cluster_name,
+            train_data=transform_output_train,
```

Comment:
I'm not sure this will work due to the way the placeholders are resolved at runtime. See my similar comment in the XGBoost sample PR.

Reply:
Thanks for noting. Yes when compilation it becomes `xgb-{{workflow.uid}}`, which is fine because Argo can capture this and replace it correctly with the workflow ID.

Reply:
Oh, I see. I confused the run id with execution id.